ethcontract/src/contract/method.rs

```diff
@@ -3,63 +3,60 @@
 //! [Instance::method](ethcontract::contract::Instance::method).
 
 use crate::transaction::{Account, GasPrice, TransactionBuilder, TransactionResult};
+use crate::{batch::CallBatch, errors::MethodError};
 use crate::{
-    batch::CallBatch,
-    errors::{revert, ExecutionError, MethodError},
+    errors::{revert, ExecutionError},
+    tokens::{Error as TokenError, SingleTokenize},
 };
 use ethcontract_common::abi::{Function, Token};
 use std::marker::PhantomData;
-use web3::contract::tokens::Detokenize;
-use web3::contract::Error as Web3ContractError;
 use web3::types::{Address, BlockId, Bytes, CallRequest, U256};
 use web3::Transport;
 use web3::{api::Web3, BatchTransport};
 
-/// A void type to represent methods with empty return types.
-///
-/// This is used to work around the fact that `(): !Detokenize`.
-pub struct Void(());
-
-/// Represents a type can detokenize a result.
-pub trait Detokenizable {
-    /// The output that this type detokenizes into.
-    type Output;
+/// A solidity return type. We need a separate trait for this so that we can represent functions
+/// that return nothing.
+pub trait Return {
+    /// The output type.
+    type Output: SingleTokenize;
```

Comment:
Does this mean we don't support `MultiTokenize` return types? ABI v2 allows returning structs for example.

Reply:
I see `SingleTokenize` supports `Tuple` as well.

Reply:
The difference is how this presented in the abi json. In ethabi the decode and encode functions work on Vec<Token>. This corresponds to function arguments being a json list and not a top level tuple, so they should encode into and from Vec<Token>. Return types are technically a list too in the json but as far as I can tell they should always be a tuple so a single Token.

Reply:
I wonder if Vec<Token> and Tuple encode to the same bytes. In that case we might be able to get rid of MultiTokenize by pretending that our arguments form a single tuple all together.

Reply:
> I wonder if Vec and Tuple encode to the same bytes

I think `Vec` encodes to `offest -> length || item0 || item1 || ...` while `Tuple` encodes to `offset -> item0 || item1 || ...`.